### PR TITLE
Formal parameter of ptr type must be instantiable

### PR DESCRIPTION
**Pull request by dneto0**
Fixes: #2268

### CODE REVIEW

File: wgsl/index.bs [hunk 1]
```diff
@@ -7344,6 +7344,9 @@ Note: The [=attribute/const=] attribute cannot be applied to user-declared funct
 * A [=formal parameter|function parameter=] [=shader-creation error|must=] one the following types:
     * a constructible type
     * a pointer type
+        * It [=shader-creation error|must=] be valid to [=variable declaration|declare a variable=]
+              whose [=reference type=] corresponds to the pointer type,
+              i.e. with the pointer type's address space, store type, and access mode.
```
Comment (jimblandy) [inline wgsl/index.bs:7349]: This works for me. You can't declare a dynamically-sized type in any of the address spaces functions accept pointers in, so this addresses my concern in #2268.

**Review by jimblandy — APPROVED** [review]
Looks good.

**Comment by dneto0** [comment]
I looked at putting this constraint at the end of the Memory View Types section https://gpuweb.github.io/gpuweb/wgsl/#memory-view-types  right after we assert the correspondence between pointers and references.  But (1) it was wordy there, and (2) I don't think the problem arises anywhere aside from function formal parameters.

Note that this PR goes a little beyond just eliminating the case of runtime-sized arrays in workgroup storage.
It also eliminates cases where the access modes are non-sensical.  Essentially, the pointer type must correspond to a variable that you *could* declare.   

I went for simpler wording where the pointer type corresponds to the *whole* variable you could declare.  I think that's general enough.  I had to think through whether buffer variables didn't have weird constraints anymore.

File: wgsl/index.bs [hunk 2]
```diff
@@ -7344,6 +7344,9 @@ Note: The [=attribute/const=] attribute cannot be applied to user-declared funct
 * A [=formal parameter|function parameter=] [=shader-creation error|must=] one the following types:
     * a constructible type
     * a pointer type
+        * It [=shader-creation error|must=] be valid to [=variable declaration|declare a variable=]
```
Comment (alan-baker) [inline wgsl/index.bs:7347]: Nit: alternative suggested wording, I have a small issue with saying corresponding reference type.

>The [=pointee type=] [shader-creation error|must=] be a valid [=store type=] for a [=variable declaration=] in the same address space.

Comment (dneto0) [reply wgsl/index.bs:7347]: I want to capture the constraints on the access mode too.  I'll edit to simplify.

**Review by alan-baker — APPROVED** [review]

**Pull request merged**